packages/@lwc/template-compiler/src/parser/index.ts

```diff
@@ -359,10 +335,62 @@ export default function parse(source: string, state: State): TemplateParseResult
         const lwcOpts = {};
         applyLwcDynamicDirective(element, lwcOpts);
         applyLwcDomDirective(element, lwcOpts);
+        applyLwcRenderModeDirective(element, lwcOpts);
+        applyLwcPreserveCommentsDirective(element, lwcOpts);
```

Comment:
ah, there is validation to make sure these are never used on a node with a parent. But you are still paying the cost of `getTemplateAttribute` for every node. It is faster to check if a node is a root or not compared to running `getTemplateAttribute` for every node.

Reply:
I have been tinkering with this code this morning to avoid invoking the `getTemplateAttribute` and it introduces more complexity than anything. The cost of running `getTemplateAttribute` is fairly low as it is just an `Array.prototype.find` in most of the cases. I would prefer to keep the code as-is for now and optimize later if necessary.